### PR TITLE
very much work-in-progress support for Microsoft SQL

### PR DESCRIPTION
**Pull request by davidpodhola**
- uses `FSharp.Data.SqlClient` for Microsoft SQL
- too much code moved to `mssql.fs` and `psql.fs` (the later is probably not working now)
- interfaces for database backend should be probably created insted of `sql.Engine`
- `validators.fs` unfinished
- `Reference` is very naively implemented as a dropdown with all the values

Generally I would like to discuss if this is the right direction :-)

### CODE REVIEW

**Comment by lefthandedgoat** [comment]
Whats with all the changes to the paket file and all the extra System.* 

Did newer versions of Npqsql add support for .net core?

File: genit.sln [hunk 1]
```diff
@@ -1,31 +1,28 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 2010
-Project("{f2a71f9b-5d33-465a-a702-920d77279786}") = "genit", "src\genit\genit.fsproj", "{64EDF01D-DC76-43CF-8BC2-5157B841510E}"
+# Visual Studio 14
```
Comment (davidpodhola) [inline genit.sln]: I am on Windows using Visual Studio 2015. This can be probably ignored.

File: paket.dependencies [hunk 2]
```diff
@@ -8,4 +8,5 @@ nuget Npgsql
 nuget Http.fs
 nuget BCrypt-Official
 nuget FsPickler
-nuget FsPickler.Json
\ No newline at end of file
+nuget FsPickler.Json
+nuget FSharp.Data.SqlClient
```
Comment (davidpodhola) [inline paket.dependencies]: See http://fsprojects.github.io/FSharp.Data.SqlClient/index.html

File: paket.lock [hunk 3]
```diff
@@ -2,21 +2,401 @@ NUGET
...
+    FsPickler.Json (2.2)
+      FsPickler (2.2)
       Newtonsoft.Json (>= 6.0.5)
     Http.fs (1.5.1)
+    Microsoft.Win32.Primitives (4.0.1-rc2-24027) - framework: >= netstandard13
+      System.Runtime (>= 4.1.0-rc2-24027) - framework: >= netstandard13
     Newtonsoft.Json (8.0.3)
-    Npgsql (3.0.5)
+    Npgsql (3.1.3)
+      Microsoft.Win32.Primitives (>= 4.0.1-rc2-24027) - framework: >= netstandard13
```
Comment (davidpodhola) [inline paket.lock]: not sure why this happened. Might be related to the .NET Core somehow?

File: src/genit/generator/generator.fs [hunk 4]
```diff
@@ -15,14 +15,6 @@ let destination filename =
   |> (fun path -> path.Replace("bin", "generated"))
   |> (fun path -> System.IO.Path.Combine(path, filename))
 
-let zipOptions (options : string list) =
```
Comment (davidpodhola) [inline src/genit/generator/generator.fs]: used in other files, so just moved

File: src/genit/generator/generator.fs [hunk 5]
```diff
@@ -1004,15 +918,15 @@ let generate (site : Site) =
   let fake_data_results = site.Pages |> List.map fakeDataTemplate |> flatten
   let generated_fake_data_result = generated_fake_data_template fake_data_results
 
-  let connectionString = sprintf "Server=127.0.0.1;User Id=%s; Password=secure123;Database=%s;" site.AsDatabase site.AsDatabase
-  let generated_data_access_result = generated_data_access_template connectionString (sql.createQueries site)
+  let connectionString = sprintf @"Data Source=(localdb)\MSSQLLocalDB;Initial Catalog=%s;Integrated Security=True" site.AsDatabase 
+  let generated_data_access_result = generated_data_access_template sql.Engine.MicrosoftSQL connectionString (sql.createQueries site sql.Engine.MicrosoftSQL)
```
Comment (davidpodhola) [inline src/genit/generator/generator.fs]: using `sql.Engine.MicrosoftSQL` now. Global function?

File: src/genit/generator/sql.fs [hunk 6]
```diff
@@ -2,74 +2,62 @@ module sql
 
 open dsl
 open helper_general
+open psql
 
-let createTemplate dbname =
-  sprintf """
-DROP DATABASE IF EXISTS %s;
-CREATE DATABASE %s;""" dbname dbname
+type Engine =
```
Comment (davidpodhola) [inline src/genit/generator/sql.fs]: use `Interface` instead.

File: src/genit/genit.fsproj [hunk 7]
```diff
@@ -78,8 +81,8 @@
     <Compile Include="generated/generated_security.fs" />
     <Compile Include="generated/generated_html.fs" />
     <Compile Include="generated/generated_types.fs" />
-    <Compile Include="generated/generated_forms.fs" />
```
Comment (davidpodhola) [inline src/genit/genit.fsproj]: the file order was intentionally changed :-)

**Comment by lefthandedgoat** [comment]
Ok I will take some time and work through this and clean it up.  Thanks!

**Comment by davidpodhola** [comment]
Yes, I guess some `.NETStandard` support was added [Npgsql](https://www.nuget.org/packages/Npgsql/). Not sure why this happened anyway.

**Comment by lefthandedgoat** [comment]
Ok I may go back to a previous version of Npgsql that doesn't have all of that stuff.  I want to avoid .netcore for now until the dust settles and the F# story is solid.

**Comment by davidpodhola** [comment]
Very true, me trying to do the same.

**Comment by lefthandedgoat** [comment]
Sorry I was very busy with work.  I have time today and my goal is to get this working!

**Comment by davidpodhola** [comment]
No problem, I can quite understand that :-) I would be very happy to commit more work to this. Still taking this PR more as the base for a discussion.

**Comment by lefthandedgoat** [comment]
Awesome.  Let me look through it all again.  It may be close enough that I can just spend an hour or two and finish it up.

**Pull request merged**